### ISSUE TITLE
(Possibly undocumented?) systemd-networkd reverse route ordering

### ISSUE THREAD

**Issue by jackprice** [systemd/systemd#5430]
**TL;DR - Routes defined in a networkd .network file seems to be applied in reverse-order, leading to errors relating to dependent routes.**

> Edit 2019-11-20 for [future travellers](https://xkcd.com/979/): The specific use case below is solved by the addition of the `GatewayOnLink` directive

---

### Submission type

  - [x] Bug report
  - [ ] Request for enhancement (RFE)

### systemd version the issue has been seen with

231 (also 219, so not a regression)

### Used distribution

Fedora Server 25 (w/ systemd-231-10.fc25)

### In case of bug report: Expected behaviour you didn't see

Gateway configured in a .network file should appear in the routing table.

### In case of bug report: Unexpected behaviour you saw

The journal shows the errors similar to

```
Feb 23 05:08:19 localhost.localdomain systemd[1]: Started Network Service.
Feb 23 05:08:19 localhost.localdomain systemd-networkd[955]: ens192: Could not set route: Network is unreachable
Feb 23 05:08:32 localhost.localdomain systemd-networkd[955]: ens192: Configured
```

### In case of bug report: Steps to reproduce the problem

1. Optional - Flush the routing table for a clean slate

2. Create a network configuration file with static routes that have dependencies between them - in my case this is the bare minimum configuration that errors (IPs partially redacted);

    ```
    [Match]
    Name=ens192

    [Address]
    Address=149.x.124.58/28

    [Route]
    Destination=149.x.124.64
    Scope=link

    [Route]
    Gateway=149.x.124.64
    ```

3. Restart systemd-networkd

4. See that the routing table contains only the route defined by the `Address` and the first static route.

    ```
    # ip route
    149.x.124.48/28 dev ens192  proto kernel  scope link  src 149.x.124.58 
    149.x.124.64 dev ens192  proto static  scope link
    ````

    The journal will show the error - `ens192: Could not set route: Network is unreachable`.

---

Executed sequentially, the routes are valid. I can't find anything in the networkd docs to suggest any kind of ordering for route configuration. 

**The error does not seem to occur when the order of the two `[Route]` sections are swapped**, which seems counterintuitive.

Is this intended functionality, but undocumented?

**Comment by ssahani** [comment]
No I can't reproduce this issue . Is there more information you can provide ?

 cat 20-en1.network
```
[Match]
Name=host

[Address]
Address=192.168.100.100/24

[Route]
Destination=192.168.100.1
Scope=link


[Route]
Gateway=192.168.100.2
```

```
[root@Zeus network]# ip route flush all
[root@Zeus network]# ip route
[root@Zeus network]# ip route
default via 192.168.100.2 dev host proto static 
192.168.100.0/24 dev host proto kernel scope link src 192.168.100.100 
192.168.100.1 dev host proto static scope link 
```



```
SYSTEMD_LOG_LEVEL=debug ./systemd-networkd
Failed to read $container of PID 1, ignoring: Permission denied
Found container virtualization none.
Sent message type=method_call sender=n/a destination=org.freedesktop.DBus object=/org/freedesktop/DBus interface=org.freedesktop.DBus member=Hello cookie=1 reply_cookie=0 error-name=n/a error-message=n/a
Got message type=method_return sender=org.freedesktop.DBus destination=:1.105 object=n/a interface=n/a member=n/a cookie=1 reply_cookie=1 error-name=n/a error-message=n/a
Sent message type=method_call sender=n/a destination=org.freedesktop.DBus object=/org/freedesktop/DBus interface=org.freedesktop.DBus member=AddMatch cookie=2 reply_cookie=0 error-name=n/a error-message=n/a
Got message type=method_return sender=org.freedesktop.DBus destination=:1.105 object=n/a interface=n/a member=n/a cookie=3 reply_cookie=2 error-name=n/a error-message=n/a
Sent message type=method_call sender=n/a destination=org.freedesktop.DBus object=/org/freedesktop/DBus interface=org.freedesktop.DBus member=RequestName cookie=3 reply_cookie=0 error-name=n/a error-message=n/a
Got message type=method_return sender=org.freedesktop.DBus destination=:1.105 object=n/a interface=n/a member=n/a cookie=5 reply_cookie=3 error-name=n/a error-message=n/a
Failed to open configuration file '/etc/systemd/networkd.conf': No such file or directory
timestamp of '/etc/systemd/network' changed
timestamp of '/usr/lib/systemd/network' changed
Ignoring /usr/lib/systemd/network/20-en1.network, because it's not a regular file with suffix .netdev.
Ignoring /usr/lib/systemd/network/tmp, because it's not a regular file with suffix .netdev.
Ignoring /usr/lib/systemd/network/tmp, because it's not a regular file with suffix .network.
host: Flags change: +UP +LOWER_UP +RUNNING +MULTICAST +BROADCAST
Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_39 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=4 reply_cookie=0 error-name=n/a error-message=n/a
host: Link 9 added
host: udev initialized link
host: Saved original MTU: 1500
guest: Flags change: +UP +LOWER_UP +RUNNING +MULTICAST +BROADCAST
Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_38 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=5 reply_cookie=0 error-name=n/a error-message=n/a
guest: Link 8 added
guest: udev initialized link
guest: Saved original MTU: 1500
virbr2-nic: Flags change: +MULTICAST +BROADCAST
virbr2-nic: Link 7 added
virbr2-nic: udev initialized link
virbr2-nic: Saved original MTU: 1500
virbr2: Flags change: +UP +MULTICAST +BROADCAST
Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_36 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=6 reply_cookie=0 error-name=n/a error-message=n/a
virbr2: Link 6 added
virbr2: udev initialized link
virbr2: Saved original MTU: 1500
virbr0-nic: Flags change: +MULTICAST +BROADCAST
virbr0-nic: Link 5 added
virbr0-nic: udev initialized link
virbr0-nic: Saved original MTU: 1500
virbr0: Flags change: +UP +MULTICAST +BROADCAST
Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_34 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=7 reply_cookie=0 error-name=n/a error-message=n/a
virbr0: Link 4 added
virbr0: udev initialized link
virbr0: Saved original MTU: 1500
wlan0: Flags change: +UP +LOWER_UP +RUNNING +MULTICAST +BROADCAST
Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_33 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=8 reply_cookie=0 error-name=n/a error-message=n/a
wlan0: Link 3 added
wlan0: udev initialized link
wlan0: Saved original MTU: 1500
eth0: Flags change: +UP +MULTICAST +BROADCAST
Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_32 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=9 reply_cookie=0 error-name=n/a error-message=n/a
eth0: Link 2 added
eth0: udev initialized link
eth0: Saved original MTU: 1500
lo: Flags change: +LOOPBACK +UP +LOWER_UP +RUNNING
Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_31 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=10 reply_cookie=0 error-name=n/a error-message=n/a
lo: Link 1 added
lo: udev initialized link
lo: Saved original MTU: 0
host: Adding address: fe80::dc06:21ff:fe7f:c639/64 (valid forever)
Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_39 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=11 reply_cookie=0 error-name=n/a error-message=n/a
host: Gained IPv6LL
guest: Adding address: fe80::a89c:43ff:fea8:9a56/64 (valid forever)
Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_38 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=12 reply_cookie=0 error-name=n/a error-message=n/a
guest: Gained IPv6LL
wlan0: Adding address: fe80::6e81:19b6:e6fb:ee41/64 (valid forever)
Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_33 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=13 reply_cookie=0 error-name=n/a error-message=n/a
wlan0: Gained IPv6LL
wlan0: Adding address: 2405:204:952f:bee4:e722:829f:fc00:b25d/64 (valid forever)
Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_33 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=14 reply_cookie=0 error-name=n/a error-message=n/a
lo: Adding address: ::1/128 (valid forever)
host: Adding address: 192.168.100.100/24 (valid forever)
Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_39 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=15 reply_cookie=0 error-name=n/a error-message=n/a
guest: Adding address: 192.168.100.100/24 (valid forever)
Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_38 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=16 reply_cookie=0 error-name=n/a error-message=n/a
virbr2: Adding address: 192.168.254.1/24 (valid forever)
virbr0: Adding address: 192.168.122.1/24 (valid forever)
wlan0: Adding address: 192.168.225.114/24 (valid for 10h 54min 46s)
lo: Adding address: 127.0.0.1/8 (valid forever)
rtnl: received address with invalid family 129, ignoring.
rtnl: received address with invalid family 128, ignoring.
Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=17 reply_cookie=0 error-name=n/a error-message=n/a
Enumeration completed
host: Link state is up-to-date
host: found matching network '/usr/lib/systemd/network/20-en1.network'
LLDP: Started LLDP client
host: Started LLDP.
host: Discovering IPv6 routers
NDISC: Started IPv6 Router Solicitation client
Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_39 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=18 reply_cookie=0 error-name=n/a error-message=n/a
Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_39 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=19 reply_cookie=0 error-name=n/a error-message=n/a
host: Setting addresses
NDISC: Sent Router Solicitation, next solicitation in 4s
guest: Link state is up-to-date
guest: Unmanaged
Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_38 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=20 reply_cookie=0 error-name=n/a error-message=n/a
virbr2-nic: Link state is up-to-date
virbr2-nic: Unmanaged
Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_37 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=21 reply_cookie=0 error-name=n/a error-message=n/a
virbr2: Link state is up-to-date
virbr2: Unmanaged
Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_36 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=22 reply_cookie=0 error-name=n/a error-message=n/a
virbr0-nic: Link state is up-to-date
virbr0-nic: Unmanaged
Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_35 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=23 reply_cookie=0 error-name=n/a error-message=n/a
virbr0: Link state is up-to-date
virbr0: Unmanaged
Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_34 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=24 reply_cookie=0 error-name=n/a error-message=n/a
wlan0: Link state is up-to-date
wlan0: Unmanaged
Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_33 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=25 reply_cookie=0 error-name=n/a error-message=n/a
eth0: Link state is up-to-date
eth0: Unmanaged
Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_32 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=26 reply_cookie=0 error-name=n/a error-message=n/a
lo: Link state is up-to-date
lo: Unmanaged
Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_31 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=27 reply_cookie=0 error-name=n/a error-message=n/a
host: Removing address: 192.168.100.100/24 (valid forever)
Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_39 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=28 reply_cookie=0 error-name=n/a error-message=n/a
ifindex=9
ifindex=9
ifindex=9
host: Adding address: 192.168.100.100/24 (valid forever)
Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_39 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=29 reply_cookie=0 error-name=n/a error-message=n/a
host: Addresses set
Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_39 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=30 reply_cookie=0 error-name=n/a error-message=n/a
host: Setting routes
host: Routes set <==============================HERE
Got message type=signal sender=org.freedesktop.DBus destination=:1.105 object=/org/freedesktop/DBus interface=org.freedesktop.DBus member=NameAcquired cookie=2 reply_cookie=0 error-name=n/a error-message=n/a
Got message type=signal sender=org.freedesktop.DBus destination=:1.105 object=/org/freedesktop/DBus interface=org.freedesktop.DBus member=NameAcquired cookie=4 reply_cookie=0 error-name=n/a error-message=n/a
NDISC: Sent Router Solicitation, next solicitation in 8s
NDISC: No RA received before link confirmation timeout
NDISC: Invoking callback for 't'.
host: Configured
Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_39 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=31 reply_cookie=0 error-name=n/a error-message=n/a
NDISC: Sent Router Solicitation, next solicitation in 16s
NDISC: Sent Router Solicitation, next solicitation in 32s
```

**Comment by Jean-Daniel** [comment]
I have the very same problem here.

```
systemd 237
+PAM +AUDIT +SELINUX +IMA +APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD -IDN2 +IDN -PCRE2 default-hierarchy=hybrid
```

```
[Match]
Name=eth0

[Network]
LinkLocalAddressing=ipv6
Address=5.39.xx.99/24
Address=2001:41d0:8:xxxx::1/128
IPv6AcceptRA=no

[Route]
Destination=0.0.0.0/0
Gateway=5.39.xx.254

[Route]
Destination=2001:41d0:8:8fff:ff:ff:ff:ff/128
Scope=link

[Route]
Destination=::/0
Gateway=2001:41d0:8:8fff:ff:ff:ff:ff
GatewayOnlink=true
```

```
systemd-networkd[11728]: eth0: Adding address: 5.39.xx.99/24 (valid forever)
systemd-networkd[11728]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/network1/link/_32 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=14 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=
systemd-networkd[11728]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/network1 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=15 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
systemd-networkd[11728]: eth0: Addresses set
systemd-networkd[11728]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/network1/link/_32 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=16 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=
systemd-networkd[11728]: eth0: Setting routes
systemd-networkd[11728]: eth0: Could not set route: No route to host
systemd-networkd[11728]: eth0: Routes set
systemd-networkd[11728]: eth0: Configured
systemd-networkd[11728]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/network1/link/_32 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=17 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=
systemd-networkd[11728]: eth0: Adding address: 2001:41d0:8:xxxx::1/128 (valid forever)
```

The scope route is properly created but the gateway route fails.

If I tri to add it afterward using the ip command, it works fine.

**Comment by Jean-Daniel** [comment]
> No I can't reproduce this issue . Is there more information you can provide ?
> 
> cat 20-en1.network
> 
> ```
> [Match]
> Name=host
> 
> [Address]
> Address=192.168.100.100/24
> 
> [Route]
> Destination=192.168.100.1
> Scope=link
> 
> 
> [Route]
> Gateway=192.168.100.2
> ```
> ```
> [root@Zeus network]# ip route flush all
> [root@Zeus network]# ip route
> [root@Zeus network]# ip route
> default via 192.168.100.2 dev host proto static 
> 192.168.100.0/24 dev host proto kernel scope link src 192.168.100.100 
> 192.168.100.1 dev host proto static scope link 
> ```

Your test case is not valid. The gateway MUST NOT be on a network directly reachable.

Try to 192.168.101.1 for your gateway address, and it should fail.

**Comment by Jean-Daniel** [comment]
Then, maybe this is an IPv6 only issue, because I'm doing this on an IPv6 link and adding GatewayOnLink does not solve the issue:

```
[Route]
Destination=2001:1234:5:8fff:ff:ff:ff:ff/128
Scope=link

[Route]
Destination=::/0
Gateway=2001:1234:5:8fff:ff:ff:ff:ff
GatewayOnlink=true
```
 
```
Nov 09 10:44:17 systemd-networkd[997]: eth0: Could not set route: No route to host
```

**Comment by jackprice** [comment]
`GatewayOnLink` was introduced in v234 (#5734), so my original issue would be solved with that addition.

I last looked into this a couple of years ago, so I don't know if it's still the case that routes are created in reverse order. If so, it should probably be documented somewhere!

**Comment by yuwata** [comment]
Fix is waiting in #10930.

**Comment by yuwata** [comment]
I noticed that the issue cannot reproducible, that is, the tests added in #10930 can pass without https://github.com/yuwata/systemd/commit/0862cc686d1740ca4fee681ad971f5ae942fe0ab. Let's close this.

**Comment by Jean-Daniel** [comment]
Does it works with IPv6 too ? I still don't managed to get it working, even with GatewayOnLink

**Comment by yuwata** [comment]
Uh, I am not sure. I will test that. Thank you for reminding me that.

**Comment by ssahani** [comment]
@Jean-Daniel could you paste the similar conf that can be reproducible.

**Comment by Jean-Daniel** [comment]
My config looks like this:

```
[Match]
Name=eth0

[Network]
LinkLocalAddressing=ipv6
Address=2001:1234:5:8f63::1/128
IPv6AcceptRA=no

[Route]
Destination=2001:1234:5:8fff:ff:ff:ff:ff/128
Scope=link

[Route]
Destination=::/0
Gateway=2001:1234:5:8fff:ff:ff:ff:ff
GatewayOnlink=true
```

Note that if I declare the Gateway route (without GatewayOnlink) before declaring the network route, it works.

**Comment by ssahani** [comment]
could you try with https://github.com/systemd/systemd/pull/10938
